src/xml_parser_main_metadata_together.c

```diff
@@ -0,0 +1,371 @@
+    return call_user_callback_if_package_finished(pkg, cb_data, err);
+}
+
+static int
+other_pkgcb(cr_Package *pkg, void *cbdata, G_GNUC_UNUSED GError **err)
+{
+    cr_CbData *cb_data = cbdata;
+    cb_data->other_count++;
+    pkg->loadingflags |= CR_PACKAGE_LOADED_OTH;
+    return call_user_callback_if_package_finished(pkg, cb_data, err);
+}
+
+static int
+primary_pkgcb(cr_Package *pkg, void *cbdata, G_GNUC_UNUSED GError **err)
+{
+    cr_CbData *cb_data = cbdata;
+    cb_data->primary_count++;
+
+    gpointer pkey = NULL;
+    gpointer pval = NULL;
+    if (g_hash_table_lookup_extended(cb_data->in_progress_pkgs, pkg->pkgId, &pkey, &pval)) {
+        // package was already encountred in some other metadata type
+
+        if (pval == NULL) {
+            //this pkg was already fully handled
+            //TODO(amatej): maybe error?
+            assert(0);
+        } else {
+            cr_package_copy_into(pkg, pval);
+            cr_package_free(pkg);
+            pkg = pval;
+        }
+    } else {
+        // we are handling never before seen package
+
+        if (cb_data->newpkgcb) {
+            // call user specified new function if provided and copy q
+            cr_Package *user_created_pkg;
+            if (cb_data->newpkgcb(&user_created_pkg, pkg->pkgId, NULL, NULL, cb_data->newpkgcb_data, err)) {
+                return CR_CB_RET_ERR;
+            } else {
+                cr_package_copy_into(pkg, user_created_pkg);
+                cr_package_free(pkg);
+                pkg = user_created_pkg;
+            }
+        }
+        g_hash_table_insert(cb_data->in_progress_pkgs, g_strdup(pkg->pkgId), pkg);
+    }
+
+    pkg->loadingflags |= CR_PACKAGE_FROM_XML;
+    pkg->loadingflags |= CR_PACKAGE_LOADED_PRI;
+
+    return call_user_callback_if_package_finished(pkg, cb_data, err);
+}
+
+static int
+parse_next_section(CR_FILE *target_file, const char *path, cr_ParserData *pd, GError **err)
+{
+    char buf[XML_BUFFER_SIZE];
+    GError *tmp_err = NULL;
+    int parsed_len = cr_read(target_file, buf, XML_BUFFER_SIZE, &tmp_err);
+    if (tmp_err) {
+        g_critical("%s: Error while reading xml '%s': %s", __func__, path, tmp_err->message);
+        g_propagate_prefixed_error(err, tmp_err, "Read error: ");
+        return -1;
+    }
+    int done = parsed_len == 0;
+    if (xmlParseChunk(pd->parser, buf, parsed_len, done)) {
+        xmlErrorPtr xml_err = xmlCtxtGetLastError(pd->parser);
+        g_critical("%s: parsing error '%s': %s", __func__, path,
+                   (xml_err) ? xml_err->message : "UNKNOWN_ERROR");
+        g_set_error(err, ERR_DOMAIN, CRE_XMLPARSER,
+                    "Parse error '%s' at line: %d (%s)",
+                    path,
+                    (xml_err) ? (int) xml_err->line : 0,
+                    (xml_err) ? (char *) xml_err->message : "UNKNOWN_ERROR");
+        return -1;
+    }
+
+    return done;
+}
+
+//TODO(amatej): add a flag that enforces pkg are in the same order (If possible?)
```

Comment:
For what it's worth, as mentioned in the email we're not aware of any tools that can generate such metadata, and Pulp 2 would not have been capable of parsing it at all. We've not had any reports of that issue, so such out-of-order metadata may not exist in the wild.  To be honest we might even consider turning on this flag.

I feel like the only obvious way it could happen is a "torn" repository with simple metadata filenames.

Reply:
I was also thinking it might be useful.